### PR TITLE
IOS7: Fix touch inputs for gamepad controller dpad events for Griffon

### PR DESCRIPTION
**Pull request by larsamannen**
The touch controls in iOS could cause  Lure of the Temptress to "hang" waiting for buttons to be "released".
The top main menu was also impossible to use in touch based systems like Android and iOS

### CODE REVIEW

**Comment by lephilousophe** [comment]
Just tested [f93cf20](https://github.com/scummvm/scummvm/pull/5187/commits/f93cf20305140695ac6a7b811c57b657ea8cd438) and it works like a charm.
LGTM.

File: backends/platform/ios7/ios7_gamepad_controller.mm [hunk 1]
```diff
@@ -162,10 +166,29 @@ - (void)controllerDidConnect:(NSNotification *)notification {
...
+					[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_LEFT isPressed:true];
+				} else if (xValue > 0) {
+					[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_RIGHT isPressed:true];
+				} else if (xValue == 0 && currentDpadXValue < 0) {
+					[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_LEFT isPressed:false];
+				} else {
+					[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_RIGHT isPressed:false];
+				}
+				currentDpadXValue = xValue;
+			} else if (yValue != currentDpadYValue) {
```
Comment (criezy) [inline backends/platform/ios7/ios7_gamepad_controller.mm:180]: With this approach you can get down events with no corresponding up events. Would that not be an issue?

Would it not be better to do something like:
```c++
// Negative values are left/down, positive are right/up, 0 is no press
// Change value to only be -1, 0, or 1
xValue = xValue < 0.f ? -1.f : xValue > 0.f ? 1.f : 0.f;
if (xValue != currentDpadXValue) {
	if (currentDpadXValue < 0) {
		[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_LEFT isPressed:false];
	} else if (currentDpadXValue > 0) {
		[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_RIGHT isPressed:false];
	}
	if (xValue < 0) {
		[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_LEFT isPressed:true];
	} else if (xValue > 0) {
		[self handleJoystickButtonAction:Common::JOYSTICK_BUTTON_DPAD_RIGHT isPressed:true];
	}
}
// And same for Y
[...]
```

File: backends/platform/ios7/ios7_gamepad_controller.mm [hunk 2]
```diff
@@ -38,6 +38,8 @@ @implementation GamepadController {
 	GCVirtualControllerConfiguration *_config;
 #endif
 #endif
+	int currentDpadXValue;
+	int currentDpadYValue;
```
Comment (criezy) [inline backends/platform/ios7/ios7_gamepad_controller.mm:42]:
```suggestion
	int _currentDpadXValue;
	int _currentDpadYValue;
```

**Review by criezy — COMMENTED** [review]
As was suggested on Discord, it could be a good idea to replace the `LURE_CLICKABLE_MENUS` define with a `kFeatureTouchScreen` feature in `OSystem`, with the Android, iOS, and Wii backend returning true for that feature.

Other engines could also benefit from knowing if the backend has a touch screen or not.

I can take a look at making that change tomorrow.

File: backends/platform/ios7/ios7_gamepad_controller.mm (hunk 1 above)
Comment (larsamannen) [reply backends/platform/ios7/ios7_gamepad_controller.mm:180]: Yes, I'll update

**Comment by sev-** [comment]
@larsamannen any updates?

**Comment by larsamannen** [comment]
> @larsamannen any updates?

I will probably split this PR and put the lure fixes in a new. I should also sync with @criezy  about the change to touchscreen feature

https://discord.com/channels/581224060529148060/711242520415174666/1137361120114524220

**Comment by larsamannen** [comment]
@criezy I've updated this PR to only have the backend fixes. I will address the Lure fixes in another PR. I've also addressed your comments.

File: backends/platform/ios7/ios7_gamepad_controller.mm (hunk 1 above)
Comment (criezy) [reply backends/platform/ios7/ios7_gamepad_controller.mm]: Was this updated? It seems to me that the issue is still present in the current code. Or with this not an issue after all?

Comment (larsamannen) [reply backends/platform/ios7/ios7_gamepad_controller.mm]: Oh, I forgot to push the last commit I had which had the changes. Sorry for that. I'll push it tomorrow.

**Review by criezy — APPROVED** [review]
Thank you. It looks good to me now.

**Pull request merged**